cmd/gcp-controller-manager/node_csr_approver.go

```diff
@@ -698,6 +698,29 @@ func clusterHasInstance(ctx *controllerContext, instanceZone string, instanceID
 		return false, err
 	}
 
+	// instanceZone looks like
+	// "https://www.googleapis.com/compute/v1/projects/my-project/zones/us-central1-c"
+	// Extract the bare zone name just "us-central1-c".
+	instanceZoneName := path.Base(instanceZone)
+
+	if ctx.csrApproverUseListReferrers {
+		klog.Infof("using ListReferrers to verify instance cluster membership")
+		referrers, err := compute.NewInstancesService(ctx.gcpCfg.Compute).ListReferrers(ctx.gcpCfg.ProjectID, instanceZoneName, fmt.Sprintf("%d", instanceID)).Do()
+		if err != nil {
+			return false, err
+		}
+
+		for _, referrer := range referrers.Items {
+			for _, ig := range clusterInstanceGroupUrls {
+				if strings.EqualFold(ig, referrer.Referrer) {
+					return true, nil
+				}
+			}
+		}
+
+		return false, nil
```

Comment:
Do you intend to fallback to validateInstanceGroupUrls, or a linear search. If so, you should not return false if verifyMembership fails.

Reply:
I mean if listReferrers fails.